src/clima_diagnostics.jl

```diff
@@ -186,13 +233,60 @@ function orchestrate_diagnostics(
         active_compute[diag_index] || continue
         diag = scheduled_diagnostics[diag_index]
 
-        diag.variable.compute!(
+        diagnostic_handler.counters[diag_index] += 1
+
+        # We have three cases:
+        #
+        # 1. The variable has an in-place `compute!` function. In this case, we simply
+        # evaluate it and overwrite the associated storage.
+        #
+        # 2a. The variable has a `compute` function that returns a Field. In this case, we
+        # copy it over to the storage.
+        #
+        # 2b. The variable has a `compute` function that returns
+        # `Base.Broadcast.Broadcasted`. In this case, we update the value in our dictionary
+        # and materialize the expression in the next loop.
+
+        has_inplace_compute = !isnothing(diag.variable.compute!)
+        if has_inplace_compute
+            # Case 1
+            diag.variable.compute!(
+                diagnostic_handler.storage[diag_index],
+                integrator.u,
+                integrator.p,
+                integrator.t,
+            )
+        else
+            out_or_broadcasted =
+                diag.variable.compute(integrator.u, integrator.p, integrator.t)
+
+            if out_or_broadcasted isa Base.Broadcast.Broadcasted
+                # Case 2b
+                diagnostic_handler.broadcasted_expressions[diag] =
+                    out_or_broadcasted
+            else
+                # Case 2a
+                diagnostic_handler.storage[diag_index] .= out_or_broadcasted
+            end
+        end
+    end
+
+    # Evaluate the lazy compute (aka, materialize everything)
+    for diag_index in 1:length(scheduled_diagnostics)
+        active_compute[diag_index] || continue
+        diag = scheduled_diagnostics[diag_index]
+        haskey(diagnostic_handler.broadcasted_expressions, diag) || continue
+
+        Base.Broadcast.materialize!(
```

Comment:
If we can check `diagnostic_handler.storage` above, then we could use `diag.variable.compute(integrator.u, integrator.p, integrator.t)` on the RHS here (`compute` will just return a pointer, and `materialize!` does the work)